README.md

# Coercible errors

Zero-cost error handling for generic traits.

## Rationale

Assume we want to build a crate that defines a generic trait,
meant to be implemented by others.
Below is a minimalist example of such a trait:

```rust
    pub trait Producer {
        fn produce(&self) -> u16;
    }
```

Some implementations might work well with this definition,
but other implementations may encounter errors in some situations
(*e.g.* `IOError` for a file-system based implementation).
We call the former kind *infallible implementations*,
and the latter kind *fallible implementations*.

In order to support both kinds of implementations,
the methods of our trait should rather return `Result<_, _>`.
This raises the question of the error type that those results should contain.

An option is to define a dedicated error type for our crate,
and force implementers to wrap their errors into this type.

```rust
    pub trait Producer {
        fn produce(&self) -> Result<u16, MyError>;
    }
```

This works, but breaks the "zero-cost abstraction" motto for infallible implementations.
Indeed, `Result<T, MyError>` can be significantly bigger than the type `T` alone.
For example, with a simple `MyError` type defined with the
[error_chain] crate,

* `<Result<(), MyError>` is 56 bytes long (*versus* 0 bytes for `()`),
* `<Result<u16, MyError>` is 64 bytes long (*versus* 2 bytes for `u16`).

A more flexible option is to let implementers specify their own error type:

```rust
    pub trait Producer {
        type Error: Error + Send + 'static;
        fn produce(&self) -> Result<u16, Self::Error>;
    }
```

For infallible implementations,
the `Error` associated type can be set to [`never`]
or any other value-less type (typically an empty Enum).
The compiler will then optimize away this error-type from the `Result`,
effectively returning only the success-type.

We now have a real zero-cost abstraction,
where infallible implementations do not pay the toll of error handling.
On the other hand, it will be harder to work with several heterogeneous implementations of our trait.
Consider for example the following type:

```rust
    pub struct PMax<P1, P2> (P1, P2);

    impl<P1: Producer, P2: Producer> Producer for PMax<P1, P2> {
        type Error = ???; // <<<< we have a problem here
        fn produce(&self) -> Result<u16, Self::Error> { 
            Ok(self.0.produce()?.max(self.1.produce()?))
        }

    }
```

Since `P1` and `P2` may use totally unrelated error types,
we don't know which error type to use.
We could use a "chainable" error type as defined by [error_chain],
but then we would go back to using a "fat" result even when `P1` and `P2` are both infallible.

## The solution

This crate provides a solution to the problems described above.
The idea is:

* to provide zero-cost error handling for infallible implementations,
  by allowing them to use [`never`] as their error type;
* to limit heterogeneity among fallible implementations,
  by requiring them to use a dedicated error type defined by the trait designer;
* to let the compiler infer the best error type when combining several implementations.

The example above would become:

```rust
    /// a dedicated error type
    pub struct MyError { /* ... */ }

    // define appropriate types and traits
    coercible_errors!(MyError);

    pub trait Producer {
        // require that Producer::Error be either MyError or never
        type Error: CoercibleWith<MyError> + CoercibleWith<Never>;
        fn produce(&self) -> Result<u16, Self::Error>;
    }

    pub struct PMax<P1, P2> (P1, P2);
    impl<P1: Producer, P2: Producer> Producer for PMax<P1, P2> 
        // this trait bound is required to be able to use CoercedError below
        where P1::Error: CoercibleWith<P2::Error>
    {
        // compute the most appropriate Error type based on P1 and P2;
        // especially, if P1 and P2 are both infallible,
        // PMax will be infallible as well.
        type Error = CoercedError<P1::Error, P2::Error>;
        fn produce(&self) -> Result<u16, Self::Error> {
            Ok(
              // the coerced error always implements From<_>
              // for both P1::Error and P2::Error,
              // so inner errors can simply be lifted with '?'
              self.0.produce()?
              .max(self.1.produce()?)
            )
        }

    }
```

The `coercible_errors` macro takes care of defining the following traits and types:

* `CoercibleWith<E>` is a trait to let the compiler infer the correct coercing of error types.
  The macro provides implementations so that [`never`] and [`never`] coerce into [`never`], and that any other combination of [`never`] and `MyError` coerce into `MyError`.
* `CoercedError<E1, E2>` is a type alias using `CoercibleWith` to determine the appropriate coerced error type.
* `CoercedResult<T, E1, E2>` is a shortcut for  `Result<T, CoercedError<E1, E2>>`.


### About [`never`]

Since the [`never`] type is currently unstable,
this crate actually defines its own version called `coercible_errors::Never`.
Once [`never`] becomes stable,
`coercible_errors::Never` will become a simple alias to [`never`],
avoiding a breaking change.


[error_chain]: https://docs.rs/error-chain/
[`never`]: https://doc.rust-lang.org/std/primitive.never.html

## License

[CECILL-C]

(compatible with GNU LGPL)

[CECILL-C]: http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html


CHANGELOG.txt
0.1.4
    Changed license to CECILL-C (french variant of LGPL)
    and improved metadata in Cargo.toml

0.1.3
    CoerceWith<Never> should also be explicit (updated macro and examples).

0.1.2
    Changed license to CECILL-B (french variant of BSD)

0.1.1
    Updated metadata and README for a better rendering on crates.io

0.1.0
    Initial release

src/test.rs
mod error {
    error_chain! {
        errors {
            Producer {
                description("error occuredn in producer"),
            }
            Consumer {
                description("error occurent in consumer"),
            }
        }
    }
    coercible_errors! {}

}

use self::error::*;
use std::result::Result as StdResult;

pub trait Producer {
    /// The error type that this producer can raise.
    /// Must be either [`Error`] or [`Never`].
    ///
    /// [`Error`]: struct.Error.html
    /// [`Never`]: enum.Never.html
    type Error: CoercibleWith<Error> + CoercibleWith<Never>;

    /// Produce a value.
    fn produce(&self) -> StdResult<u16, Self::Error>;
}

impl Producer for u16 {
    type Error = Never;
    fn produce(&self) -> StdResult<u16, Self::Error> {
        Ok(*self)
    }
}

impl Producer for u32 {
    type Error = Error;
    fn produce(&self) -> StdResult<u16, Self::Error> {
        if *self <= 0xFFFF {
            Ok(*self as u16)
        } else {
            bail!("Value too big to be produced")
        }
    }
}

pub trait Consumer {
    type Error: CoercibleWith<Error> + CoercibleWith<Never>;
    fn consume(&mut self, val: u16) -> StdResult<(), Self::Error>;
}

impl Consumer for u16 {
    /// The error type that this producer can raise.
    /// Must be either [`Error`] or [`Never`].
    ///
    /// [`Error`]: struct.Error.html
    /// [`Never`]: enum.Never.html
    type Error = Never;

    /// Consume a value.
    fn consume(&mut self, val: u16) -> StdResult<(), Self::Error> {
        *self = val;
        Ok(())
    }
}

impl Consumer for u8 {
    type Error = Error;
    fn consume(&mut self, val: u16) -> StdResult<(), Self::Error> {
        if val <= 0xff {
            *self = val as u8;
            Ok(())
        } else {
            bail!("Value too big to be consumed")
        }
    }
}

pub struct PMax<P1, P2>(P1, P2);

impl<P1: Producer, P2: Producer> Producer for PMax<P1, P2>
where
    P1::Error: CoercibleWith<P2::Error>,
{
    type Error = CoercedError<P1::Error, P2::Error>;
    fn produce(&self) -> CoercedResult<u16, P1::Error, P2::Error> {
        Ok(self.0.produce()?.max(self.1.produce()?))
    }
}

/// This is the naive version of pipe;
/// it always returns a `Result`,
/// even if both the producer and the consumer return `OkResult`s.
fn pipe1<P: Producer, C: Consumer>(p: &P, c: &mut C) -> Result<()> {
    let v = p.produce().chain_err(|| ErrorKind::Producer)?;
    c.consume(v).chain_err(|| ErrorKind::Consumer)
}

/// This is the smart version of pipe;
/// it returns the smallest possible result (`Result` or `OkResult`),
/// based on what the producer and the consumer return.
fn pipe2<P: Producer, C: Consumer>(p: &P, c: &mut C) -> CoercedResult<(), P::Error, C::Error>
where
    P::Error: CoercibleWith<C::Error>,
{
    Ok(c.consume(p.produce()?)?)
}

#[test]
fn test() -> Result<()> {
    // NB: most of this test is actually performed at compile time:
    // we check that the coerced result types are as expected
    // (Result<_> or OkResult<_>).

    let mut cons8: u8 = 0;
    let mut cons16: u16 = 0;

    // ########## pipe1 ##########
    // pipe1 always chain inner errors into Error
    let _r: Result<()> = pipe1(&42_u16, &mut cons16);
    let _r: Result<()> = pipe1(&42_u16, &mut cons8);
    let _r: Result<()> = pipe1(&42_u32, &mut cons16);
    let _r: Result<()> = pipe1(&42_u32, &mut cons8);

    let _r: Result<()> = pipe1(&0x20000_u32, &mut cons8);
    let _r: Result<()> = pipe1(&0x200_u16, &mut cons8);

    // this is already a good thing,
    // because only methods needing to *coerce* errors
    // need to fallback to Error;
    // simple methods may still use Self::Error, e.g.:
    let _r: OkResult<u16> = 42_u16.produce();
    let _r: OkResult<()>  = cons16.consume(42);
    let _r: Result<u16>   = 42_u32.produce();
    let _r: Result<()>    = cons8.consume(42);

    // ########## pipe2 ##########
    // pipe2 infers the minimal type from its arguments
    let _r: OkResult<()> = pipe2(&42_u16, &mut cons16);
    let _r: Result<()>   = pipe2(&42_u16, &mut cons8);
    let _r: Result<()>   = pipe2(&42_u32, &mut cons16);
    let _r: Result<()>   = pipe2(&42_u32, &mut cons8);

    let _r: Result<()>   = pipe2(&0x20000_u32, &mut cons8);
    let _r: Result<()>   = pipe2(&0x200_u16, &mut cons8);

    // ######## PMax ########
    let _r: OkResult<u16> = PMax(42_u16, 1_u16).produce();
    let _r: Result<u16>   = PMax(42_u32, 1_u16).produce();
    let _r: Result<u16>   = PMax(42_u16, 1_u32).produce();
    let _r: Result<u16>   = PMax(42_u32, 1_u32).produce();

    // ######## testing the returned values ########
    // (having the correct type is not enough...)

    let r1: Result<()> = pipe1(&42_u16, &mut cons16);
    let r2: Result<()> = pipe1(&42_u16, &mut cons8);
    let r3: Result<()> = pipe1(&42_u32, &mut cons16);
    let r4: Result<()> = pipe1(&42_u32, &mut cons8);
    //println!("{:?} {:?} {:?} {:?}", r1, r2, r3, r4);
    assert!(r1.is_ok());
    assert!(r2.is_ok());
    assert!(r3.is_ok());
    assert!(r4.is_ok());

    let r5: Result<()> = pipe1(&0x20000_u32, &mut cons8);
    let r6: Result<()> = pipe1(&0x200_u16, &mut cons8);
    //println!("{:?}\n{:?}", r5, r6);
    assert!(r5.is_err());
    assert!(r6.is_err());

    let r1: OkResult<()> = pipe2(&42_u16, &mut cons16);
    let r2: Result<()>   = pipe2(&42_u16, &mut cons8);
    let r3: Result<()>   = pipe2(&42_u32, &mut cons16);
    let r4: Result<()>   = pipe2(&42_u32, &mut cons8);
    //println!("{:?} {:?} {:?} {:?}", r1, r2, r3, r4);
    assert!(r1.is_ok());
    assert!(r2.is_ok());
    assert!(r3.is_ok());
    assert!(r4.is_ok());

    let r5: Result<()> = pipe2(&0x20000_u32, &mut cons8);
    let r6: Result<()> = pipe2(&0x200_u16, &mut cons8);
    //println!("{:?}\n{:?}", r5, r6);
    assert!(r5.is_err());
    assert!(r6.is_err());

    let r1: OkResult<u16> = PMax(42_u16, 1_u16).produce();
    let r2: Result<u16>   = PMax(42_u32, 1_u16).produce();
    let r3: Result<u16>   = PMax(42_u16, 1_u32).produce();
    let r4: Result<u16>   = PMax(42_u32, 1_u32).produce();
    //println!("{:?} {:?} {:?} {:?}", r1, r2, r3, r4);
    assert!(r1.unwrap() == 42);
    assert!(r2.unwrap() == 42);
    assert!(r3.unwrap() == 42);
    assert!(r4.unwrap() == 42);

    println!("All tests passed",);

    Ok(())
}


src/lib.rs
//! Zero-cost error handling for generic traits.
//!
//! # Rationale
//!
//! Assume you want to build a crate that defines a generic trait,
//! meant to be implemented by others.
//! Some implementations of that trait will always succeed,
//! others may sometimes fail.
//! The methods of the generic trait should therefore return `Result<_,_>`,
//! but that should not induce an overhead for infallible implementations
//! (per the *zero-cost abstraction* motto).
//!
//! The [`coercible_errors!`] macro will define a set of traits and types
//! that you can use to define your generic traits,
//! in order to keep their error-handling as efficient as possible.
//! More precisely, the compiler will be able to optimize away the error types
//! whenever only infallible implementations of the generic trait are used.
//!
//! See `README.md` for a more detailed explaination.
//!
//! [`coercible_errors!`]: macro.coercible_errors.html

/// Sets up coercible_errors for a previously defined error type.
///
/// It re-exports the types [`Never`] and [`OkResult`],
/// and defines three new traits and types `CoercibleWith`,
/// `CoercedError` and `CoercedResult`.
///
/// [`Never`]: enum.Never.html
/// [`OkResult`]: type.OkResult.html
#[macro_export]
macro_rules! coercible_errors {
    () => {
        coercible_errors!(Error);
    };
    ($error: ty) => {
        coercible_errors!($error, CoercibleWith, CoercedError, CoercedResult);
    };
    ($error: ty, $coercible_with: ident, $coerced_error: ident, $coerced_result: ident) => {
        pub use $crate::{Never, OkResult};

        // This conversion can never happen (since Never can have no value),
        // but it is required for allowing $error and Never to coerce with each other.
        impl From<Never> for $error {
            fn from(_: Never) -> $error {
                unreachable!()
            }
        }

        /// A trait used to determine how to best coerce two error types.
        ///
        /// In practice, the only two error types that it handles are `$error` or `Never`.
        pub trait $coercible_with<E>:
            Sized + std::marker::Send + std::error::Error + 'static
        {
            type Into: std::marker::Send
                + std::error::Error
                + 'static
                + From<Self>
                + From<E>
                + $coercible_with<$error>
                + $coercible_with<Never>;
        }
        impl $coercible_with<$error> for $error {
            type Into = $error;
        }
        impl $coercible_with<Never> for $error {
            type Into = $error;
        }
        impl $coercible_with<$error> for Never {
            type Into = $error;
        }
        impl $coercible_with<Never> for Never {
            type Into = Never;
        }

        /// A shortcut for building the coerced error type,
        /// given two error types,
        /// which must both be either `$error` or `Never`.
        pub type $coerced_error<E1, E2> = <E1 as $coercible_with<E2>>::Into;

        /// A shortcut for building the coerced result type,
        /// given one value type and two error types,
        /// which must both be either `$error` or `Never`.
        pub type $coerced_result<T, E1, E2> = std::result::Result<T, $coerced_error<E1, E2>>;
    };
}

/// An "error" type that can never happen.
///
/// NB: once the [`never`] types reaches *stable*,
/// this type will be an alias for the standard type.
///
/// [`never`]: https://doc.rust-lang.org/std/primitive.never.html
///
#[derive(Clone, Debug)]
pub enum Never {}
impl ::std::fmt::Display for Never {
    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
        write!(f, "Never")
    }
}
impl std::error::Error for Never {}

/// Type alias for a result that will Never fail.
pub type OkResult<T> = std::result::Result<T, Never>;

#[cfg(feature = "example_generated")]
pub mod example_generated;

#[cfg(test)]
#[macro_use]
extern crate error_chain;

#[cfg(test)]
mod test;


src/main.rs
//! This crate is an exploration for building
//! a zero-cost error-handling for generic type.
//! 
//! The idea is that:
//! the methods of a generic type should return Results,
//! to account for implementation that may fail.
//! But *some* implementations may never fail,
//! and those would be penalized
//! by the overhead of wrapping their returned values into Results.
//! 
//! To avoid that, teach the compiler to merge two errors into the minimal supertype.
//! 
//! The scaffolding is entierly deterministic,
//! so it could be provided by a simple macro.
//! 
//! The boilerplate to be included in "smart functions"
//! (those returning a "merged result")
//! could possibly also be generated, by a procedural macro,
//! but that's beyond my macro skills for the moment.
//! 
#[macro_use] extern crate error_chain;
#[macro_use] extern crate mergeable_errors;

mod error {
    error_chain!{
        errors {
            Producer {
                description("error occuredn in producer"),
            }
            Consumer {
                description("error occurent in consumer"),
            }
        }
    }
    mergeable_errors!{}

}

use self::error::*;
use std::result::Result as StdResult;


pub trait Producer {
    /// The error type that this producer can raise.
    /// Must be either [`Error`] or [`Never`].
    /// 
    /// [`Error`]: struct.Error.html
    /// [`Never`]: enum.Never.html
    type Error: MergesWith<Error>;

    /// Produce a value.
    fn produce(&self) -> StdResult<u16, Self::Error>;
}

impl Producer for u16 {
    type Error = Never;
    fn produce(&self) -> StdResult<u16, Self::Error> { Ok(*self) }
}

impl Producer for u32 {
    type Error = Error;
    fn produce(&self) -> StdResult<u16, Self::Error> {
        if *self <= 0xFFFF {
            Ok(*self as u16)
        } else {
            bail!("Value too big to be produced")
        }
    }
}

pub trait Consumer {
    type Error: MergesWith<Error>;
    fn consume(&mut self, val: u16) -> StdResult<(), Self::Error>;
}

impl Consumer for u16 {
    /// The error type that this producer can raise.
    /// Must be either [`Error`] or [`Never`].
    /// 
    /// [`Error`]: struct.Error.html
    /// [`Never`]: enum.Never.html
    type Error = Never;

    /// Consume a value.
    fn consume(&mut self, val: u16) -> StdResult<(), Self::Error> {
        *self = val;
        Ok(())
    }
}

impl Consumer for u8 {
    type Error = Error;
    fn consume(&mut self, val: u16) -> StdResult<(), Self::Error> {
        if val <= 0xff {
            *self = val as u8;
            Ok(())
        } else {
            bail!("Value too big to be consumed")
        }
    }
}

pub struct PMax<P1, P2> (P1, P2);

impl<P1: Producer, P2: Producer> Producer for PMax<P1, P2>
    where P1::Error: MergesWith<P2::Error>
{
    type Error = MergedError<P1::Error, P2::Error>;
    fn produce(&self) -> MergedResult<u16, P1::Error, P2::Error> {
        Ok(self.0.produce()?.max(self.1.produce()?))
    }

}

/// This is the naive version of pipe;
/// it always returns a `Result`,
/// even if both the producer and the consumer return `OkResult`s.
fn pipe1<P: Producer, C: Consumer>(p: &P, c: &mut C)-> Result<()> {
    let v = p.produce().chain_err(|| ErrorKind::Producer)?;
    c.consume(v).chain_err(|| ErrorKind::Consumer)
}

/// This is the smart version of pipe;
/// it returns the smallest possible result (`Result` or `OkResult`),
/// based on what the producer and the consumer return.
fn pipe2<P: Producer, C: Consumer>(p: &P, c: &mut C)
    -> MergedResult<(), P::Error, C::Error>
where
    P::Error: MergesWith<C::Error>,
{
    Ok(c.consume(p.produce()?)?)
}


fn main() -> Result<()> {
    println!("Result<u16>  : {} bytes", std::mem::size_of::<Result<u16>>());
    println!("OkResult<u16>: {} bytes", std::mem::size_of::<OkResult<u16>>());
    println!("Result<()>   : {} bytes", std::mem::size_of::<Result<()>>());
    println!("OkResult<()> : {} bytes", std::mem::size_of::<OkResult<()>>());

    let mut cons8: u8 = 0;
    let mut cons16: u16 = 0;

    // ########## pipe1 ##########
    // pipe1 always chain inner errors into Error
    let _r: Result<()> = pipe1(&42_u16, &mut cons16);
    let _r: Result<()> = pipe1(&42_u16, &mut cons8);
    let _r: Result<()> = pipe1(&42_u32, &mut cons16);
    let _r: Result<()> = pipe1(&42_u32, &mut cons8);

    let _r: Result<()> = pipe1(&0x20000_u32, &mut cons8);
    let _r: Result<()> = pipe1(&0x200_u16,   &mut cons8);

    // this is already a good thing,
    // because only methods needing to *merge* errors
    // need to fallback to Error;
    // simple methods may still use Self::Error, e.g.:
    let _r: OkResult<u16> = 42_u16.produce();
    let _r: OkResult<()> = cons16.consume(42);
    let _r: Result<u16> = 42_u32.produce();
    let _r: Result<()> = cons8.consume(42);

    // ########## pipe2 ##########
    // pipe2 infers the minimal type from its arguments
    let _r: OkResult<()> = pipe2(&42_u16, &mut cons16);
    let _r: Result<()>   = pipe2(&42_u16, &mut cons8);
    let _r: Result<()>   = pipe2(&42_u32, &mut cons16);
    let _r: Result<()>   = pipe2(&42_u32, &mut cons8);


    let _r: Result<()>   = pipe2(&0x20000_u32, &mut cons8);
    let _r: Result<()>   = pipe2(&0x200_u16,   &mut cons8);

    // ######## PMax ########
    let _r: OkResult<u16> = PMax(42_u16, 1_u16).produce();
    let _r: Result<u16>   = PMax(42_u32, 1_u16).produce();
    let _r: Result<u16>   = PMax(42_u16, 1_u32).produce();
    let _r: Result<u16>   = PMax(42_u32, 1_u32).produce();


    // ######## testing the returned values ########
    // (having the correct type is not enough...)

    let r1: Result<()> = pipe1(&42_u16, &mut cons16);
    let r2: Result<()> = pipe1(&42_u16, &mut cons8);
    let r3: Result<()> = pipe1(&42_u32, &mut cons16);
    let r4: Result<()> = pipe1(&42_u32, &mut cons8);
    //println!("{:?} {:?} {:?} {:?}", r1, r2, r3, r4);
    assert!(r1.is_ok());
    assert!(r2.is_ok());
    assert!(r3.is_ok());
    assert!(r4.is_ok());

    let r5: Result<()> = pipe1(&0x20000_u32, &mut cons8);
    let r6: Result<()> = pipe1(&0x200_u16,   &mut cons8);
    //println!("{:?}\n{:?}", r5, r6);
    assert!(r5.is_err());
    assert!(r6.is_err());

    let r1: OkResult<()> = pipe2(&42_u16, &mut cons16);
    let r2: Result<()>   = pipe2(&42_u16, &mut cons8);
    let r3: Result<()>   = pipe2(&42_u32, &mut cons16);
    let r4: Result<()>   = pipe2(&42_u32, &mut cons8);
    //println!("{:?} {:?} {:?} {:?}", r1, r2, r3, r4);
    assert!(r1.is_ok());
    assert!(r2.is_ok());
    assert!(r3.is_ok());
    assert!(r4.is_ok());

    let r5: Result<()> = pipe2(&0x20000_u32, &mut cons8);
    let r6: Result<()> = pipe2(&0x200_u16,   &mut cons8);
    //println!("{:?}\n{:?}", r5, r6);
    assert!(r5.is_err());
    assert!(r6.is_err());

    let r1: OkResult<u16> = PMax(42_u16, 1_u16).produce();
    let r2: Result<u16>   = PMax(42_u32, 1_u16).produce();
    let r3: Result<u16>   = PMax(42_u16, 1_u32).produce();
    let r4: Result<u16>   = PMax(42_u32, 1_u32).produce();
    //println!("{:?} {:?} {:?} {:?}", r1, r2, r3, r4);
    assert!(r1.unwrap() == 42);
    assert!(r2.unwrap() == 42);
    assert!(r3.unwrap() == 42);
    assert!(r4.unwrap() == 42);


    println!("All tests passed", );

    Ok(())
}
